pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -594,6 +597,29 @@ impl LayerInner {
         }
     }
 
+    /// Use our weak ref to Timeline to build a local file path: this will
+    /// fail if our parent Timeline has been destroyed: layers should not
+    /// do any I/O after this point, and therefore not require knowledge of path.
+    pub(crate) fn local_path(&self) -> Result<Utf8PathBuf, DownloadError> {
+        let timeline = self
+            .timeline
+            .upgrade()
+            .ok_or_else(|| DownloadError::TimelineShutdown)?;
+
+        Ok(self.build_local_path(&timeline.tenant_shard_id, &timeline.timeline_id))
+    }
+
+    /// Use this instead of `local_path` if you already have a Timeline to hand.
+    pub(crate) fn build_local_path(
```

Comment:
Who uses this from outside?

Reply:
Used by:
- Timeline, constructing paths for layer files in e.g. create_delta_layer
- RemoteTimelineClient, constructing path to download to.